File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        turtoise = head
        rabbit = head

        while rabbit!=None and rabbit.next!=None:
            turtoise=turtoise.next
            rabbit = rabbit.next.next

            if rabbit == turtoise:
                return True
        return False
        

File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode tort = head;
        ListNode rabbit = head;

        while (tort != null && rabbit.next != null){
            tort = tort.next;
            rabbit = rabbit.next.next;
            if (rabbit == tort) {
                return true;
            }
        }
        return false;

        

    }
}